### PR TITLE
ARROW-9183: [C++] Fix build with clang & old libstdc++.

### PR DESCRIPTION
**Pull request by hcoona**
See https://issues.apache.org/jira/browse/ARROW-9183# for further details & discussion.

### CODE REVIEW

File: cpp/src/arrow/util/atomic_shared_ptr.h [hunk 1]
```diff
@@ -24,7 +24,9 @@
 namespace arrow {
 namespace internal {
 
-#if !defined(__clang__) && defined(__GNUC__) && __GNUC__ < 5
+// https://gcc.gnu.org/onlinedocs/libstdc++/manual/abi.html#abi.versioning
+// https://gcc.gnu.org/develop.html#timeline
+#if defined(__GLIBCXX__) && __GLIBCXX__ < 20150422
 
 // atomic shared_ptr operations only appeared in gcc 5,
 // emulate them with unsafe ops on gcc 4.x.
```
Comment (bkietz) [inline cpp/src/arrow/util/atomic_shared_ptr.h:32]:
```suggestion
#if defined(__GLIBCXX__) && __GLIBCXX__ < 20150422

// Atomic shared_ptr operations only appeared in gcc 5's libstdc++,
// emulate them with unsafe ops if unavailable.
// 
// The libstdc++ version is a the encoded release date of gcc 5, see
// https://gcc.gnu.org/onlinedocs/libstdc++/manual/abi.html#abi.versioning.__GLIBCXX__
```

**Review by bkietz — CHANGES_REQUESTED** [review]
This looks correct, but I'd like the comment to be a little more informative for the next reader

**Review by bkietz — APPROVED** [review]
LGTM, thanks

**Comment by bkietz** [comment]
Hmm, there's a failure building with GCC 4.8 https://github.com/apache/arrow/pull/7493/checks?check_run_id=791725319#step:9:534
The `#ifdef` condition seems to be failing to detect too-old libstdc++

**Review by bkietz — CHANGES_REQUESTED** [review]
@hcoona please resolve the CI failures

**Comment by hcoona** [comment]
Got it. It seems that the gcc 4.8.5 is released later than gcc 4.9.2 & gcc 5.1. I'll take a future investigation on how to detect the missing of `std::atomic_load` overloads.

**Review by bkietz — APPROVED** [review]
LGTM

🚀 sfinae, nice

**Pull request closed without merging**